cloud-controller-manager/do/firewall_controller.go

```diff
@@ -372,3 +493,30 @@ func (fc *firewallCache) updateCache(currentFirewall *godo.Firewall) {
 	defer fc.mu.Unlock()
 	fc.firewall = currentFirewall
 }
+
+func (fc *FirewallController) observeReconcileDuration(ctx context.Context, origin string) error {
+	var labels prometheus.Labels
+	switch origin {
+	case originEventAdd, originEventUpdate, originEventDelete, originFirewallLoop:
+		labels = prometheus.Labels{"reconcile_type": origin}
+	default:
+		labels = prometheus.Labels{"reconcile_type": "none"}
+	}
+	t := prometheus.NewTimer(fc.fwManager.metrics.reconcileDuration.With(labels))
+	defer t.ObserveDuration()
+
+	return fc.ensureReconciledFirewall(ctx)
+}
+
+func (fc *FirewallController) observeRunLoopDuration(ctx context.Context, stopCh <-chan struct{}, fwReconcileFreq time.Duration) error {
+	labels := prometheus.Labels{"success": "false"}
+	t := prometheus.NewTimer(fc.fwManager.metrics.runLoopDuration.With(labels))
+	defer t.ObserveDuration()
+
+	err := fc.actualRun(stopCh, fwReconcileFreq)
+	if err == nil {
+		labels["success"] = "true"
```

Comment:
So I dug down into the implementation of `With` just now and see that [the values are being extracted from the given `prometheus.Labels` at the time that `With` is called](https://github.com/prometheus/client_golang/blob/4675060215e9866ac3e834cfd62ba3fb36b89e12/prometheus/vec.go#L191).

Because of this I think you can wrap this `With` call with a `prometheus.ObserverFunc` the way you are doing above to properly capture the value of labels at the time the deferred `t.ObserveDuration()` call is run. Sorry again for the churn!